Add Cargo package autodiscovery

loispostula opened the pull request:
Implement cargo package auto-discovery

Dependson:

- [x] Support autodiscovery of private crates
- [x] Support autodiscovery of crate specifying their version inline

### Support autodiscovery of crate specifying their version inline

Cargo dependencies can be specified in two ways:
```toml
[dependencies]
rand = "0.1.0"
``` 
or 
```toml
[dependencies]
[dependencies.rand]
version = "0.1.0"
```


Due to limitation in the current `toml` parsing capabilities, it's not possible to perform such complex queries on the `Cargo.toml ` files. This means that autodiscovery will not work for dependencies not specifies using the full spec

## Test

To test this pull request, you can run the following commands:

```shell
cd pkg/plugins/autodiscovery/cargo
go test
```

## Additional Information

### Tradeoff

### Potential improvement

* Allow to restrict version update to a specific type such as only major/minor/patch update e if semver is used
* Target executing ``cargo generate-lockfile` in dry-run mode shouldn't not change the file `Cargo.lock`
* Leverage [google/osv.dev](https://github.com/google/osv.dev) to detect CVEs

olblak commented:
> Due to limitation in the current toml parsing capabilities, it's not possible to perform such complex queries on the Cargo.toml files. This means that autodiscovery will not work for dependencies not specifies using the full spec

Do you have example of queries that do no work?
Because the two example that you provide are working.

```
sources:
  example1:
    kind: toml
    spec:
      file: '/tmp/example2.toml'
      key: 'dependencies.rand'

  example2:
    kind: toml
    spec:
      file: '/tmp/example.toml'
      key: 'dependencies.rand.version'
```

loispostula commented on pkg/core/engine/main.go:
```diff
@@ -504,9 +504,9 @@ func (e *Engine) LoadAutoDiscovery(defaultEnabled bool) error {
-				manifest.SCMs[p.Config.Spec.AutoDiscovery.ScmId] = *sc
+			manifest.SCMs = make(map[string]scm.Config)
+			for scmId, sc := range p.SCMs {
+				manifest.SCMs[scmId] = *sc.Config
```
@olblak This is the biggest change in this pr in term of complete workflow, as instead of only adding the initial manifest scm to the generate manifests, I forward all scm spec to the generated manifests

olblak replied:
Interesting, I am curious what's the motivation?

loispostula replied:
The recomended way to parse cargo registries is to pull git repositories containing the index. (In https://github.com/updatecli/updatecli/pull/1081 I've implemented API support because the official cargo registry is big and the clone operation takes about 1 minute).

This means that if we want to run autodiscovery in a git repo (leveraging the current scmid process of autodiscovery), I cannot use a cargo registry in a git scm without this changes

olblak replied:
Thanks, the change is legit and I can't think of a limitation excepted being sure that there is no collision between the scmid used by the registry versus the one specified in the autodiscovery manifest

olblak commented on e2e/updatecli.d/warning.d/autodiscovery/cargo/example.1.yaml:
```diff
@@ -0,0 +1,23 @@
+  default:
+    kind: git
+    spec:
+      url: https://github.com/loispostula/cargo-lab.git
```
I am starting to wonder if we shouldn't create a new Github organization named `updatecli-test` to store those kind of repsitories.

loispostula replied:
That would work well I think

olblak replied:
I just created https://github.com/orgs/updatecli-test 
That would be great if you could migrate the repository over there as it will be more maintainable

olblak commented on pkg/plugins/autodiscovery/cargo/dependencies.go:
```diff
@@ -0,0 +1,207 @@
+		RegistryAuthToken          string
+		RegistryHeaderFormat       string
+	}{
+		ManifestName:               fmt.Sprintf("Bump %s %q for \"%s\" crate", dependencyType, dependency.Name, crateName),
```
```suggestion
		ManifestName:               fmt.Sprintf("Bump %s %q for %q crate", dependencyType, dependency.Name, crateName),
```

%q surround the string value with quotes

olblak commented on pkg/plugins/autodiscovery/cargo/dependencyManifest.go:
```diff
@@ -0,0 +1,59 @@
+
+const (
+	// dependencyManifest is the Go template used to generate
+	// the Helm chart manifests specific for Helm dependencies
```
```suggestion
	// the manifests to update the cargo file
```

olblak commented on pkg/plugins/autodiscovery/cargo/dependencyManifest.go:
```diff
@@ -0,0 +1,59 @@
+{{- end }}
+    spec:
+      file: '{{ .File }}'
+      Query: '{{ .ConditionQuery }}'
```
```suggestion
      query: '{{ .ConditionQuery }}'
```

keys should all be lowercase

olblak replied:
while the uppercase key should work. The jsonschema generated would display would consider it as an error

olblak commented on pkg/plugins/autodiscovery/cargo/utils.go:
```diff
@@ -0,0 +1,118 @@
+	"github.com/updatecli/updatecli/pkg/plugins/utils/dasel"
+)
+
+// searchChartFiles search, recursively, for every files named Cargo.toml from a root directory.
```
```suggestion
// findCargoFiles search, recursively, for every files named Cargo.toml from a root directory.
```

olblak commented on pkg/plugins/autodiscovery/cargo/utils.go:
```diff
@@ -0,0 +1,118 @@
+
+// searchChartFiles search, recursively, for every files named Cargo.toml from a root directory.
+func findCargoFiles(rootDir string, validFiles []string) ([]string, error) {
+	manifestsFiles := []string{}
```
```suggestion
	cargoFiles := []string{}
```

olblak replied:
Maybe just nitpicking

olblak replied:
Then you would have to change all occurrence of `manifestsFile` to `cargoFiles`

olblak commented on pkg/plugins/autodiscovery/cargo/utils.go:
```diff
@@ -0,0 +1,118 @@
+
+	logrus.Debugf("%d chart(s) found", len(manifestsFiles))
+	for _, foundFile := range manifestsFiles {
+		chartName := filepath.Base(filepath.Dir(foundFile))
```
```suggestion
		cargoFile := filepath.Base(filepath.Dir(foundFile))
```

olblak commented on pkg/plugins/autodiscovery/cargo/utils.go:
```diff
@@ -0,0 +1,118 @@
+	logrus.Debugf("%d chart(s) found", len(manifestsFiles))
+	for _, foundFile := range manifestsFiles {
+		chartName := filepath.Base(filepath.Dir(foundFile))
+		logrus.Debugf("    * %q", chartName)
```
```suggestion
		logrus.Debugf("    * %q", cargoFile)
```

olblak commented on pkg/plugins/resources/cargopackage/spec.go:
```diff
@@ -1,18 +1,15 @@
 type Spec struct {
-	// [S][C] IndexUrl specifies the url of the index to use to check version
-	// If set, package version will be checked against the API
-	IndexUrl string `yaml:",omitempty"`
```
This is a breaking change that could easily be avoided with a deprecation approach
We try to carefully introduce new "user" facing parameters as we want to avoid as much as possible breaking changes.

Something like 

```
`IndexUrl string `yaml:",omitempty" jsonschema:"-"``
```
Means that the filed won't appear in the generated jsonschema and so will return an error if the indexurl parameter is used.

Then in the new object you could have something like 

```
if spec.IndexURL != "" {
  switch spec.Registry.URL != {
    case true:
      logrus.Warningf("Registry.URL and IndexURL are mutually exclusive, unset indexurl")
      spec.IndexURL = ""
    case false:
      logrus.Warningf("indexurl is deprecated in favor of registry.url")
      spec.IndexURL = ""
  }
}

By emptying the deprecated field in favor of the new field, the following command will automatically update the old manifest

`updatecli manifest upgrade`

olblak reviewed (changes requested):
I made some comments and spot some breaking changes that should be avoided

loispostula commented on pkg/plugins/autodiscovery/cargo/main.go:
```diff
@@ -0,0 +1,71 @@
+	// Only specify required rule to restrict Cargo.toml update.
+	Only MatchingRules `yaml:",omitempty"`
+	// Auths provides a map of registry credentials where the key is the registry URL without scheme
+	Registries map[string]cargo.Registry `yaml:",omitempty"`
```
Yes it is used here: https://github.com/updatecli/updatecli/blob/cargo-autodiscovery/pkg/plugins/autodiscovery/cargo/dependencies.go#L59-L63

This is the way to specify custom registries (private) for crates

loispostula commented:
I've update according to the review, I think the remaining items are:

- [ ] Is it ok to provide all scmid
- [ ] Should we move the test repos to an `updateclitest` org
- [ ] Deprecation and `cargo manifest upgrade` changes. I've implemented those in the New resource code.

olblak commented on pkg/plugins/utils/cargo/main.go:
```diff
@@ -0,0 +1,19 @@
+	HeaderFormat string `yaml:"headerFormat,omitempty"`
+}
+
+type Registry struct {
```
I like this approach but maybe, we miss a validation function that check that we didn't specified at the same time an URL, a rootdir, or a scmid.

loispostula replied:
Would that be in a `func (s Spec) Validate() (bool)` function?

olblak replied:
I think I would create a function 
`func (r Registry) Validate() (bool)` here
Then call that function in the `New` function

olblak commented:
Congrats on the pullrequest, I tried it on `github.com/kubewarden/kwctl` and I only faced one issue.
It couldn't handle this line https://github.com/kubewarden/kwctl/blob/ad831c472db07385e304e0338b159741da5b9a1d/Cargo.toml#L25

We should find a way to skip dependencies that can't be handle such as if version is empty then skip the dependency.

Regarding the failing dependency, we could generate an traditional type of manifest that rely on a gittag for the source and the target would have to update the tag.
That being said, it's ok to handle this scenario later on, in a separated pullrequest. 

I think the current pullrequest face the same issue than my PR on https://github.com/updatecli/updatecli/pull/1065#issuecomment-1370197660
It needs a way to update the file `Cargo.lock`
This will probably be simplified once my shell refactoring is merged

loispostula commented:
Thanks for the extensive review, I had not foreseen the `Cargo.lock` challenge, I've added the mechanism to generate it, in the same way as in https://github.com/updatecli/updatecli/pull/1065 .

Regarding the issue for the git dependency, I've also added a placeholder for git and local path dependency and am currently ignore them.

olblak commented:
> Thanks for the extensive review, I had not foreseen the `Cargo.lock` challenge, I've added the mechanism to generate it, in the same way as in #1065 .

I didn't foreseen neither on the NPM pullrequest until very late.
I need to finalize https://github.com/updatecli/updatecli/pull/1071 which shouldn't take long now  

> Regarding the issue for the git dependency, I've also added a placeholder for git and local path dependency and am currently ignore them.

Make sense once thing at a time

loispostula commented:
@olblak Thanks for #1071 and #1067 

I've added thoses key to the manifest, I also add to make sure that the shell resource inherited the path environment variable. And it's now working fine

olblak commented:
I'll run some tests over the weekend

olblak commented on pkg/plugins/autodiscovery/cargo/dependencyManifest.go:
```diff
@@ -0,0 +1,81 @@
+    sourceid: '{{ .ExistingSourceID }}'
+targets:
+  {{ .TargetID }}:
+    name: 'Bump crate dependency "{{ .DependencyName }}" for crate "{{ .CrateName }}"'
```
Having the version information in the title would be usefull when creating the pullrequest.
Something like 

```suggestion
    name: 'Bump crate dependency "{{ .DependencyName }}" to {{ source ".SourceID" }}"'
```

`{{ source ".SourceID" }}` must be correctly escaped as it needs to be update at run time.
In the NPM pullrequest, I generated the name from the golang code instead of the go template to make it more readable

olblak commented on pkg/plugins/autodiscovery/cargo/dependencyManifest.go:
```diff
@@ -0,0 +1,81 @@
+    scmid: '{{ .ScmID }}'
+{{ end }}
+    spec:
+      command: cargo generate-lockfile
```
Is there a way to run this command in dry-run mode?

olblak replied:
Maybe something like this  cargo update --dry-run $DRY_RUN <package_name>

olblak replied:
You initial approach by using `cargo generate-lockfile` seems to be the best one.
My only drawback is when running updatecli in diff mode but I don't think there is anything we could do :(

olblak commented on pkg/plugins/autodiscovery/cargo/dependencyManifest.go:
```diff
@@ -0,0 +1,81 @@
+{{- end }}
+    spec:
+      file: '{{ .TargetFile }}'
+      key: '{{ .TargetKey }}'
```
An issue I faced on the npm pullrequest is the situation where I want the dependency "~4.0". The lockfile contains "4.0.0" but "4.0.10" is available. I don't want to update the manifest from "~4.0" to "4.0.10", I only update the lockfile.

I am wondering how it works with cargo

olblak replied:
It appears you initial approach was already good

olblak commented on pkg/plugins/utils/cargo/main.go:
```diff
@@ -0,0 +1,59 @@
+	// [A][S][C] Token specifies the cargo registry token to use for authentication.
+	Token string `yaml:",omitempty"`
+	// [A][S][C] HeaderFormat specifies the cargo registry header format to use for authentication (defaults to `Bearer`).
+	HeaderFormat string `yaml:"headerFormat,omitempty"`
```
```suggestion
	HeaderFormat string `yaml:"headerformat,omitempty"`
```
All user facing configuration must be lowercase. It's a jsonschema constraint otherwise the IDE returns an error.

olblak commented:
@loispostula In my last commit, I disable updating the Cargo.toml if it's not strict semver. So 1.0.0 should be updated to >=1.0.0 while "1.0" should only update "Cargo.lock" by running `cargo generate-lockfile`

I have mixed feeling about overidding version constraint in Cargo.toml (and I am open for discussion)
